File: python/canvas.py
import numpy as np

from color import Color


class Canvas:
    def __init__(
        self, width: int, height: int, background_color: Color | None = None
    ) -> None:
        self.__pixels = np.zeros((width, height, 3), dtype=np.uint8)
        self.__width = width
        self.__height = height
        self.__background_color = background_color

    @property
    def width(self) -> int:
        return self.__width

    @property
    def height(self) -> int:
        return self.__height

    @property
    def sizes(self) -> tuple[int, int]:
        return self.__width, self.__height

    def set_pixel(self, x: int, y: int, color: Color):
        self.__pixels[x, y] = color.R, color.G, color.B

    def get_pixel(self, x: int, y: int) -> Color:
        # return self.__data.get((x, y), self.__background_color)
        return Color(*self.__pixels[x, y])

    def get(self):
        return self.__pixels

File: src/Canvas.cpp
#include "Canvas.hpp"

void accumulate_frame(Canvas &target, const Canvas &source, int frame_num) {
    int width = target.get_width();
    int height = target.get_height();

#pragma omp parallel for schedule(dynamic)
    for (int row = 0; row < height; ++row) {
        for (int col = 0; col < width; ++col) {
            Color prev_color = target.get_pixel(row, col);
            Color curr_color = source.get_pixel(row, col);
            Color updt_color = (prev_color * (frame_num - 1) + curr_color) / frame_num;
            target.set_pixel(row, col, updt_color);
        }
    }
}